smqtk_descriptors/impls/descriptor_generator/pytorch.py

```diff
@@ -0,0 +1,538 @@
+            self.img_mat_list = img_mat_list
+            self.transform = transform
+
+        def __getitem__(self, index: int) -> "torch.Tensor":
+            return self.transform(self.img_mat_list[index])
+
+        def __len__(self) -> int:
+            return len(self.img_mat_list)
+except NameError:
+    pass
+
+
+class TorchModuleDescriptorGenerator (DescriptorGenerator):
+    """
+    Descriptor generator using some torch module.
+
+    :param smqtk.algorithms.ImageReader image_reader:
+        Image reader algorithm to use for image loading.
+    :param int|None image_load_threads:
+        Number of threads to use for parallel image loading. Set to "None" to
+        use all available CPU threads. This is 1 (serial) by default.
+    :param None|str weights_filepath:
+        Optional filepath to saved network weights to load. If this value is
+        None then implementations should attempt to use their "pre-trained"
+        mode if they have one. Otherwise an error is raised on model load when
+        not provided.
+    :param int|None image_tform_threads:
+        Number of threads to utilize when transforming image matrices for
+        network input. Set to `None` to use all available CPU threads.
+        This is 1 (serial) by default).
+    :param int batch_size:
+        Batch size use for model computation.
+    :param bool use_gpu:
+        If the model should be loaded onto, and processed on, the GPU.
+    :param None|int cuda_device:
+        Specific device value to pass to the CUDA transfer calls if `use_gpu`
+        us enabled.
+    :param None|int|float|str normalize:
+        Optionally normalize descriptor vectors as they are produced. We use
+        ``numpy.linalg.norm`` so any valid value for the ``ord`` parameter is
+        acceptable here.
+    :param bool iter_runtime:
+        By default the input image matrix data is accumulated into a list in
+        order to utilize `torch.utils.data.DataLoader` for batching.
+        If this is parameter is True however, we use a custom parallel
+        iteration pipeline for image transformation into network appropriate
+        tensors.
+        Using this mode changes the maximum RAM usage profile to be linear with
+        batch-size instead of input image data amount, as well as having lower
+        latency to first yield.
+        This mode is idea for streaming input or high input volume situations.
+        This mode currently has a short-coming of working only in a threaded
+        manner so it is not as fast as using the `DataLoader` avenue.
+    :param bool global_average_pool:
+        Optionally apply a GAP operation to the spatial dimension of the feature
+        vector that is returned by the descriptor generator. Some models return
+        a w x h x k tensor and flatten them into a single dimension, which can
+        remove the spatial context of the features. Taking an average over each
+        channel can improve the clustering in some cases.
```

Comment:
We can remove the parameter types here as the type annotations should supersede these.